pkg/retriever/retrieve.go

```diff
@@ -235,43 +252,133 @@ func runRetrievalCandidate(ctx context.Context, cfg *RetrievalConfig, client Ret
+		queryCandidate := &queryCandidate{queryResponse, candidate, time.Since(phaseStartTime)}
+		retrieval.queue.Put(queryCandidate)
+	}
+
+	if err != nil {
+		retrieval.sendEvent(eventpublisher.Failure(phaseStartTime, eventpublisher.QueryPhase, candidate, err.Error()))
+		if ctx.Err() != nil { // cancelled, don't report the error
+			retrieval.sendResult(retrievalResult{PhaseStart: phaseStartTime, PeerID: candidate.MinerPeer.ID})
+		} else {
+			retrieval.sendResult(retrievalResult{PhaseStart: phaseStartTime, PeerID: candidate.MinerPeer.ID, Err: err})
+		}
+
+		// Reduce attempts since we won't be making a retrieval for it
+		attempts := <-retrieval.attemptsChan
+		attempts--
+		retrieval.attemptsChan <- attempts
+	}
+}
+
+func queryPhase(
+	ctx context.Context,
+	cfg *RetrievalConfig,
+	client RetrievalClient,
+	timeout time.Duration,
+	candidate types.RetrievalCandidate,
+	onConnected func(),
+) (*retrievalmarket.QueryResponse, error) {
+
+	queryCtx := ctx // separate context so we can capture cancellation vs timeout
+
+	if timeout != 0 {
+		var timeoutFunc func()
+		queryCtx, timeoutFunc = context.WithDeadline(ctx, time.Now().Add(timeout))
+		defer timeoutFunc()
+	}
+
+	queryResponse, err := client.RetrievalQueryToPeer(queryCtx, candidate.MinerPeer, candidate.RootCid, onConnected)
+	if err != nil {
+		if ctx.Err() == nil { // not cancelled, maybe timed out though
+			log.Warnf(
+				"Failed to query miner %s for %s: %v",
+				candidate.MinerPeer.ID,
+				candidate.RootCid,
+				err,
+			)
+			return nil, fmt.Errorf("%w: %v", ErrQueryFailed, err)
+		}
+		// don't register an error on cancel, this is normal on a successful retrieval on an alternative SP
+		return nil, nil
+	}
+
+	return queryResponse, nil
+}
+
+func runRetrievalPhase(
+	ctx context.Context,
+	cfg *RetrievalConfig,
+	client RetrievalClient,
+	retrieval *retrieval,
+) {
+	retrieveSem := make(chan struct{}, 1) // only allow one retriever at a time
+	success := make(chan struct{}, 1)     // success signal
+
+	phaseStartTime := time.Now()
+
+	retrieveSem <- struct{}{}
+	for {
+		// Exit routine on success
+		select {
+		case <-success:
+			return
+		default:
+		}
+
+		// check retrieval attempts, stop looping if we're out of attempts
+		attempts := <-retrieval.attemptsChan
+		if attempts <= 0 {
+			retrieval.attemptsChan <- attempts
+			break
+		}
+		retrieval.attemptsChan <- attempts
 
-		if retrieval.canSendResult() { // move on to retrieval phase
-			// only one goroutine is allowed to execute past here at a time, so retrieval.retrievalStartTime
-			// doesn't need to be protected by a mutex, but it should mark the time the first retrieval
-			// starts
-			if retrieval.retrievalStartTime.IsZero() {
-				retrieval.retrievalStartTime = time.Now()
+		// retrieval routine
+		go func() {
+			// timeout the queue after a short time to not block forever,
+			// closing the routine and checking for success or no more attempts
+			getCtx, cancelCtx := context.WithTimeout(ctx, time.Second)
```

Comment:
Having this block here really isn't all that workable. Even the tests alone take five seconds to run because of it. You're essentially handling the case where:
- when you first start waiting, there are on going query attempts 
- those query attempts fail though, and then you're out of attempts with no new retrievals to attempt

The real problem becomes clear when you spell out what this case is:
- I the retriever, am ready for my next retrieval attempt
- I will wait until:
   - a retrieval is available to attempt *or*
   - there are no retrievals left to attempt
   
So there are really two signals, not one. I can imagine some sort of select here maybe?

Reply:
Correct, the problem I ran into was that queries would fail and this would be stuck waiting with no way to unblock. Internally, the queue is waiting to receive on it's internal itemsChan. Oh! Idea. We could have it also check if it's empty internally and return something like `T, bool`. We could check for empty, and return true if it's empty, false otherwise. Then we just check for an `empty` value and release the resource if false.

```golang
queryCandidate, empty := retrieval.queue.Get()
if empty {
  <-retrieveSem
  return
}
```